extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,211 @@
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          Pair key = new Pair(keyLabelPrefix + DEFAULT_KEY_STRING,
+                            keyRow.getEvent().entrySet().stream().findFirst().get().getValue());
+          if (!mergeList.contains(key)) {
+            mergeList.add(key);
+          }
+        }
+      }
+      catch (IOException e) {
+        throw e;
+      }
+      catch (ClassCastException e) {
+        log.error(e, "Encountered ClassCastException exception, please use parsers that return MapBasedInputRow based rows.");
+      }
+    }
+
+    if (valueParser != null) {
+      CloseableIterator<InputRow> iterator = valueParser.read();
+
+      return new CloseableIterator<InputRow>()
+      {
+        @Override
+        public boolean hasNext()
+        {
+          return iterator.hasNext();
+        }
+
+        @Override
+        public InputRow next()
+        {
+          InputRow row = null;
+          if (!iterator.hasNext()) {
+            throw new NoSuchElementException();
+          }
+          try {
+            // Return type for the value parser should be of type MapBasedInputRow
+            // Parsers returning other types are not compatible currently.
+            MapBasedInputRow valueRow = (MapBasedInputRow) iterator.next();
+            Map<String, Object> event = new HashMap<>();
+            HashSet<String> newDimensions = new HashSet<String>(valueRow.getDimensions());
+            for (Pair<String, Object> ele : mergeList) {
+              event.put(ele.lhs, ele.rhs);
+              newDimensions.add(ele.lhs);
+            }
+            /* Currently we prefer payload attributes if there is a collision in names.
+                We can change this beahvior in later changes with a config knob. This default
+                behavior lets easy porting of existing inputFormats to the new one without any changes.
+              */
+            event.putAll(valueRow.getEvent());
+
+            // Remove the dummy timestamp added in KafkaInputFormat
+            newDimensions.remove("__kif_auto_timestamp");
+
+            final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+            final List<String> dimensions;
+            if (!schemaDimensions.isEmpty()) {
+              dimensions = schemaDimensions;
+            } else {
+              dimensions = Lists.newArrayList(
+                  Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+              );
+            }
+            row = new MapBasedInputRow(
+                inputRowSchema.getTimestampSpec().extractTimestamp(event),
+                dimensions,
+                event
+            );
+          }
+          catch (ClassCastException e) {
+            log.error(e, "Encountered ClassCastException exception, please use parsers that return MapBasedInputRow based rows.");
```

Comment:
you want to throw an IOException here and abort processing, otherwise you'd log a message for every single record, and in addition return a null InputRow for every record, which might cause issues further down.